cpp/src/io/utilities/datasource.cpp

```diff
@@ -127,19 +170,58 @@ class memory_mapped_source : public datasource {
   size_t map_offset_ = 0;
 };
 
-std::unique_ptr<datasource> datasource::create(const std::string filepath,
+/**
+ * @brief Wrapper class for user implemented data sources
+ *
+ * Holds the user-implemented object with a non-owning pointer; The user object is not deleted
+ * when the wrapper object is destroyed.
+ * All API calls are forwarded to the user datasource object.
+ */
+class user_datasource_wrapper : public datasource {
+ public:
+  explicit user_datasource_wrapper(datasource *const source) : source(source) {}
+
+  size_t host_read(size_t offset, size_t size, uint8_t *dst) override
+  {
+    return source->host_read(offset, size, dst);
+  }
+
+  std::unique_ptr<buffer> host_read(size_t offset, size_t size) override
+  {
+    return source->host_read(offset, size);
+  }
+
+  bool supports_device_read() const override { return source->supports_device_read(); }
+
+  size_t device_read(size_t offset, size_t size, uint8_t *dst) override
+  {
+    return source->device_read(offset, size, dst);
+  }
+
+  std::unique_ptr<buffer> device_read(size_t offset, size_t size) override
+  {
+    return source->device_read(offset, size);
+  }
+
+  size_t size() const override { return source->size(); }
+
+ private:
+  datasource *const source;
```

Comment:
How is the lifetime of the datasource managed? Is a smart pointer needed?

Reply:
The wrapper does not manage the lifetime of the source object (hence raw pointer). This is the only reason we have this wrapper - internally we pass a unique_ptr<datasource> because we create the datasource object ourselves. The custom datasource is the exception, where user creates the datasource and might want to reuse it for multiple calls. I documented this in the class docs above. I'll also add a comment here.

The other option is to use a `shared_ptr` to pass the datasource to libcudf. In this case we can modify the `create` function to return `shared_ptr` too, so we wouldn't need this wrapper.

I went with this option because it is consistent with the `data_sink` side. If others agree, I think the source/sink part of the API could be improved in a separate PR (in a release after 0.14).

Reply:
added the comment